File: Data Structures & Algorithms/binary-tree-diameter/submission-2.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @return {number}
     */
    diameterOfBinaryTree(root) {
        // our counter
        const diameter = [0];

        // get the depth using dfs
        this.dfs(root, diameter);

        return diameter[0];
    }

    dfs(root, diameter){
        if (root === null){
            return 0;
        }

        let leftTree = this.dfs(root.left, diameter);
        let rightTree = this.dfs(root.right, diameter);

        // diameter = left + right
        // we want the largest diameter
        // do Max, cause we dont want to over write the diameter with smaller values
        diameter[0] = Math.max(diameter[0], leftTree + rightTree);

        // return height to parent
        return 1 + Math.max(leftTree, rightTree);
    }
}

File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        
        def dfs(root):
            if not root: 
                return False

            if root.val == subRoot.val:
                if sameTree(root, subRoot):
                    return True

            return dfs(root.right) or dfs(root.left)


        def sameTree(p, q):
            if not p and not q:
                return True

            if not p or not q or p.val != q.val:
                return False

            return sameTree(p.right, q.right) and sameTree(p.left, q.left)

        return dfs(root)